File: RxTools-library/src/main/java/com/vondear/rxtools/view/progressing/animation/interpolator/KeyFrameInterpolator.java
package com.vondear.rxtools.view.progressing.animation.interpolator;

import android.animation.TimeInterpolator;
import android.view.animation.Interpolator;

/**
 * @author vondear
 */
public class KeyFrameInterpolator implements Interpolator {

    private TimeInterpolator interpolator;
    private float[] fractions;


    public KeyFrameInterpolator(TimeInterpolator interpolator, float... fractions) {
        this.interpolator = interpolator;
        this.fractions = fractions;
    }

    public static KeyFrameInterpolator easeInOut(float... fractions) {
        KeyFrameInterpolator interpolator = new KeyFrameInterpolator(Ease.inOut());
        interpolator.setFractions(fractions);
        return interpolator;
    }

    public static KeyFrameInterpolator pathInterpolator(float controlX1, float controlY1,
                                                        float controlX2, float controlY2,
                                                        float... fractions) {
        KeyFrameInterpolator interpolator = new KeyFrameInterpolator(PathInterpolatorCompat.create(controlX1, controlY1, controlX2, controlY2));
        interpolator.setFractions(fractions);
        return interpolator;
    }

    public void setFractions(float... fractions) {
        this.fractions = fractions;
    }

    @Override
    public synchronized float getInterpolation(float input) {
        if (fractions.length > 1) {
            for (int i = 0; i < fractions.length - 1; i++) {
                float start = fractions[i];
                float end = fractions[i + 1];
                float duration = end - start;
                if (input >= start && input <= end) {
                    input = (input - start) / duration;
                    return start + (interpolator.getInterpolation(input)
                            * duration);
                }
            }
        }
        return interpolator.getInterpolation(input);
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/progressing/animation/interpolator/KeyFrameInterpolator.kt
package com.tamsiree.rxui.view.progressing.animation.interpolator

import android.animation.TimeInterpolator
import android.view.animation.Interpolator
import com.tamsiree.rxui.view.progressing.animation.interpolator.Ease.inOut

/**
 * @author tamsiree
 */
class KeyFrameInterpolator(private val interpolator: TimeInterpolator, private vararg var fractions: Float) : Interpolator {
    fun setFractions(vararg fractions: Float) {
        this.fractions = fractions
    }

    @Synchronized
    override fun getInterpolation(input: Float): Float {
        var input = input
        if (fractions.size > 1) {
            for (i in 0 until fractions.size - 1) {
                val start = fractions[i]
                val end = fractions[i + 1]
                val duration = end - start
                if (input >= start && input <= end) {
                    input = (input - start) / duration
                    return start + (interpolator.getInterpolation(input)
                            * duration)
                }
            }
        }
        return interpolator.getInterpolation(input)
    }

    companion object {
        @JvmStatic
        fun easeInOut(vararg fractions: Float): KeyFrameInterpolator {
            val interpolator = KeyFrameInterpolator(inOut())
            interpolator.setFractions(*fractions)
            return interpolator
        }

        @JvmStatic
        fun pathInterpolator(controlX1: Float, controlY1: Float,
                             controlX2: Float, controlY2: Float,
                             vararg fractions: Float): KeyFrameInterpolator {
            val interpolator = KeyFrameInterpolator(PathInterpolatorCompat.create(controlX1, controlY1, controlX2, controlY2))
            interpolator.setFractions(*fractions)
            return interpolator
        }
    }

}